### PR TITLE
 Functional test

### PR DESCRIPTION
**Pull request by qifeng-bai**
It includes functional tests and some small bug fixes

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -39,6 +39,7 @@
     "angular-ui-bootstrap": "^1.3.3",
     "bootbox": "^5.0.1",
     "bootstrap": "3.4.1",
+    "chromedriver": "89.0.0",
```
Comment (adam-collins) [inline package.json:42]: not test only?

**Review by adam-collins — CHANGES_REQUESTED** [review]
Changes required. 

Can we have a smaller koppen.kml and koppen.zip?

File: grails-app/services/au/org/ala/spatial/portal/HubWebService.groovy [hunk 2]
```diff
@@ -137,51 +161,95 @@ class HubWebService {
...
                 }
+
                 if (entity) {
                     ((EntityEnclosingMethod) call).setRequestEntity(entity)
                 }
             }
 
+            /**
+             * HttpHeaders.COOKIE, 'ALA-Auth=   will trigger CAS redirect to login page.
+             * Disabling CORS may break it
```
[the remark this answers is not in the document]
Comment (qifeng-bai) [reply grails-app/services/au/org/ala/spatial/portal/HubWebService.groovy:247]: Spatial-service relies on apiKey + userId to validate users, like what Biocollect does

File: grails-app/controllers/au/org/ala/spatial/portal/LogController.groovy [hunk 3]
```diff
@@ -10,28 +10,26 @@ class LogController {
...
+        def r = hubWebService.urlResponse(HttpPost.METHOD_NAME, url, null, headers,
+                new StringRequestEntity(request.JSON.toString()), true)
         render status: r.statusCode
     }
 
     def search() {
         String url = "${grailsApplication.config.layersService.url}/log/search"
         def headers = [:]
-        headers.put("apiKey",grailsApplication.config.api_key)
-        headers.put("userId", authService.userId)
```
[the remark this answers is not in the document]
Comment (qifeng-bai) [reply grails-app/controllers/au/org/ala/spatial/portal/LogController.groovy:28]: pass authtication=true to urlResponse, and then urlResponse will insert apikey and userId

File: grails-app/controllers/au/org/ala/spatial/portal/LogController.groovy [hunk 4]
```diff
@@ -10,28 +10,26 @@ class LogController {
...
 
+    /**
+     * Post log
+     * @return
+     */
     def index() {
         String url = "${grailsApplication.config.layersService.url}/log/"
         def headers = [:]
-        headers.put("apiKey",grailsApplication.config.api_key)
-        headers.put("userId", authService.userId)
```
[the remark this answers is not in the document]
Comment (qifeng-bai) [reply grails-app/controllers/au/org/ala/spatial/portal/LogController.groovy:17]: urlResponse will insert apikey and userId when authentication is true

File: grails-app/assets/javascripts/spApp/templates/legendContent.tpl.htm [hunk 5]
```diff
@@ -89,7 +89,8 @@
                     {{x.displayName}}
                 </option>
 
-                <option ng-repeat="x in selected.layer.groupedFacets" value="{{x.facet}}" ng-disabled="x.separator">
+                <option ng-repeat="x in selected.layer.groupedFacets" value="{{x.facet}}" ng-disabled="x.separator" 5
```
[the remark this answers is not in the document]
Comment (qifeng-bai) [reply grails-app/assets/javascripts/spApp/templates/legendContent.tpl.htm:92]: removed

File: grails-app/services/au/org/ala/spatial/portal/HubWebService.groovy [hunk 6]
```diff
@@ -204,19 +272,25 @@ class HubWebService {
...
-            def responseBody = call.responseBody
+            def data
             try {
                 def ct = responseHeaders['Content-Type']
                 if (ct && !ct.toString().startsWith('image') &&
                         (ct.toString().startsWith("text") || ct.toString().startsWith("application/json"))) {
-                    responseBody = new String(responseBody, 'UTF-8')
+
+                    def is
+                    if (responseHeaders["Content-Encoding"] == "gzip") {
```
[the remark this answers is not in the document]
Comment (qifeng-bai) [reply grails-app/services/au/org/ala/spatial/portal/HubWebService.groovy:282]: When CAS intervene SS by wrong config,  it sends gzip data back. The previous code tries to read it as text and then forward unrecognised text to SP.  This fix is used to discover what content is.  It is never used after CAS config fixed

File: package.json (hunk 1 above)
Comment (qifeng-bai) [reply package.json:42]: We don't have chromedriver.89 in grail repo.  Have to install it via npm and then link to Grails repo

File: grails-app/assets/javascripts/spApp/controller/areaReportCtrl.js [hunk 7]
```diff
@@ -407,27 +410,29 @@
...
+                                }
                             } else {
-                                $scope.count(items[index], BiocacheService.speciesCount(items[index].query, items[index].extraQ)).then(function (count) {
-                                    $scope.countNextItem(items, index + 1)
-                                })
+                                $scope.countNextItem(items, index + 1)
                             }
                         } else {
-                            $scope.countNextItem(items, index + 1)
+                            $scope.makeCSV()
```
[the remark this answers is not in the document]
Comment (qifeng-bai) [reply grails-app/assets/javascripts/spApp/controller/areaReportCtrl.js:427]: Done. Move it to parent if

**Comment by qifeng-bai** [comment]
> Changes required.
> 
> Can we have a smaller koppen.kml and koppen.zip?

koppen.kml is 180k, and koppen.zip is 2.3M.  They are the smallest shape files I can find

**Review by adam-collins — APPROVED** [review]
#409

**Pull request merged**